### PR TITLE
feat/script strategy

### PR DESCRIPTION
**Pull request by aarmoa**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
This PR creates a simplified version of the trading strategies called Script. The scripts are intended to be used without requiring a configuration. All data used by the script should be included in the script file.
The script is a Python class. It can be created by subclassing the new ScriptStrategyBase class.
All scripts modules should be stored in the `/scripts` folder.
The start command has been modified to allow the execution of scripts:
`start --script <script_module_name>`

**Tests performed by the developer**:
All unit tests passing.
Executed a client running the DCA example script with binance paper trade connector.

**Tips for QA testing**:
Test the example scripts in the scripts folder.
Please validate the docker image can be created and the scripts run correctly in it.

This PR comes from https://github.com/CoinAlpha/hummingbot/pull/109

### CODE REVIEW

**Comment by PtrckM** [comment]
- unable to start one of the sample script strategy `'buy_dip_example'` 

```
2022-04-05 14:30:55,172 - 31960 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: 'await' outside async function (buy_dip_example.py, line 95)
Traceback (most recent call last):
  File "/Users/patrick/Documents/HB/191/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/Users/patrick/Documents/HB/191/hummingbot/client/command/start_command.py", line 109, in start_check
    await self.start_market_making(self.strategy_name, restore)
  File "/Users/patrick/Documents/HB/191/hummingbot/client/command/start_command.py", line 127, in start_market_making
    self.start_script_strategy()
  File "/Users/patrick/Documents/HB/191/hummingbot/client/command/start_command.py", line 112, in start_script_strategy
    script_strategy = ScriptStrategyBase.load_script_class(self.strategy_file_name)
  File "/Users/patrick/Documents/HB/191/hummingbot/strategy/script_strategy_base.py", line 58, in load_script_class
    script_module = importlib.import_module(f".{script_name}", package=SCRIPT_STRATEGIES_MODULE)
  File "/opt/anaconda3/envs/hummingbot/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 779, in exec_module
  File "<frozen importlib._bootstrap_external>", line 916, in get_code
  File "<frozen importlib._bootstrap_external>", line 846, in source_to_code
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/patrick/Documents/HB/191/scripts/buy_dip_example.py", line 95
    proposal: List[OrderCandidate] = await self.create_proposal()
                                     ^
SyntaxError: 'await' outside async function
2022-04-05 14:30:58,024 - 31960 - hummingbot.core.rate_oracle.rate_oracle - INFO - Network status has changed to NetworkStatus.CONNECTED. Starting networking...
```

image

**Comment by PtrckM** [comment]
- getting new error

image

```
2022-04-06 05:33:51,000 - 23995 - hummingbot.core.clock - ERROR - Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 119, in hummingbot.core.clock.Clock.run_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/strategy_py_base.pyx", line 37, in hummingbot.strategy.strategy_py_base.StrategyPyBase.c_tick
    self.tick(timestamp)
  File "/home/ubuntu/scriptstra/hummingbot/strategy/script_strategy_base.py", line 80, in tick
    self.on_tick()
  File "/home/ubuntu/scriptstra/scripts/buy_dip_example.py", line 96, in on_tick
    proposal: List[OrderCandidate] = self.create_proposal()
  File "/home/ubuntu/scriptstra/scripts/buy_dip_example.py", line 105, in create_proposal
    daily_closes = asyncio.get_event_loop().run_until_complete(self.get_daily_close_list(self.trading_pair))
  File "/home/ubuntu/miniconda3/envs/scriptstra/lib/python3.8/asyncio/base_events.py", line 592, in run_until_complete
    self._check_running()
  File "/home/ubuntu/miniconda3/envs/scriptstra/lib/python3.8/asyncio/base_events.py", line 552, in _check_running
    raise RuntimeError('This event loop is already running')
RuntimeError: This event loop is already running

```

logs:
[buy_dip.zip](https://github.com/CoinAlpha/hummingbot/files/8423983/buy_dip.zip)

**Review by PtrckM — APPROVED** [review]
LGTM!
Test summary:

- Install and compile feature branch without any issues.
- Start script strategy using the sample scripts provided and confirmed no issues.
- Modify the scripts to run on live without issues.
- Connect api keys, confirmed exchange balances are reflecting correct
- Started the strategy without issues, confirmed orders are created based on config
- Ran a test bot for several hours without errors
- Confirmed in logs that there are no error showing when orders are filled
- Trade information are correct (price, size) based on history command output
- Compared trades in client vs exchange matched.
- Compare fees vs csv vs exchange matched.
- No issue building docker image 
- Use the create.sh docker script without issue
- Start a docker instance bot using start.sh docker script without issue.
- Copy sample scripts to docker instance bot folder hummingbot_scripts and start script strategy without any issue.

**Comment by Nullably** [comment]
Good idea to rebrand script to pmm_script since that's the only strategy it supports.
Thanks @aarmoa for taking this over and clean up a number of things.
@PtrckM I think one important feature (not mentioned) of this is to allow users make changes to the script and have it reflected on the bot without exiting the bot (though you will need to stop and start the strategy). It'd be good to have it tested. You can modify, say, `buy_interval` dca_example.py and see if the bot picks up the change.

File: test/hummingbot/strategy/test_script_strategy_base.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    end: pd.Timestamp = pd.Timestamp("2019-01-01 01:00:00", tz="UTC")
+    start_timestamp: float = start.timestamp()
+    end_timestamp: float = end.timestamp()
+    connector_name: str = "mock_paper_exchange"
+    trading_pair: str = "HBOT-USDT"
+    base_asset, quote_asset = trading_pair.split("-")
+    base_balance: int = 500
+    quote_balance: int = 5000
+    initial_mid_price: int = 100
+    clock_tick_size = 1
```
Comment (petioptrv) [inline test/hummingbot/strategy/test_script_strategy_base.py:33]: We should put those in a class setup method for uniformity with the rest of our test modules

**Review by petioptrv — CHANGES_REQUESTED** [review]

**Comment by rxlxrxsx** [comment]
- One minor issue when trying to run a non-existing script file (or accidentally entering a non-existing script file), the bot recognizes the file and still tries to start the strategy.
Screen Shot 2022-04-07 at 4 10 06 PM

```python
2022-04-07 16:06:35,772 - 5814 - root - ERROR - No module named 'hummingbot.strategy.non-existing'
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot/hummingbot/client/config/config_helpers.py", line 202, in get_strategy_starter_file
    strategy_module = __import__(f"hummingbot.strategy.{strategy}.start",
ModuleNotFoundError: No module named 'hummingbot.strategy.non-existing'
```
```python
2022-04-07 16:06:35,774 - 5814 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: 
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/Users/reiniersalazar/hummingbot/hummingbot/client/command/start_command.py", line 109, in start_check
    await self.start_market_making(self.strategy_name, restore)
  File "/Users/reiniersalazar/hummingbot/hummingbot/client/command/start_command.py", line 133, in start_market_making
    raise NotImplementedError
NotImplementedError
```

- Also for clarification, I'm trying to look for an example on how to properly run the `pmm_script` command but can't seem to find any.
Screen Shot 2022-04-07 at 4 41 23 PM

File: hummingbot/client/command/start_command.py [hunk 2]
```diff
@@ -50,15 +42,17 @@ async def wait_till_ready(self,  # type: HummingbotApplication
...
             self.ev_loop.call_soon_threadsafe(self.start, log_level, restore)
             return
-        safe_ensure_future(self.start_check(log_level, restore), loop=self.ev_loop)
+        safe_ensure_future(self.start_check(log_level, restore, script), loop=self.ev_loop)
 
     async def start_check(self,  # type: HummingbotApplication
                           log_level: Optional[str] = None,
-                          restore: Optional[bool] = False):
+                          restore: Optional[bool] = False,
+                          strategy_file_name: Optional[str] = None):
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply hummingbot/client/command/start_command.py:55]: My opinion is that in the code we should treat scripts as if they were normal strategies. That is why I would vote for keeping the name "strategy" for variables and parameters.

**Comment by aarmoa** [comment]
Hello @rxlxrxsx. PMM Scripts are not meant to be executed by themselves. They can only be configured to run in parallel to a PMM Strategy. Check the global configuration to find out how to enable PMM script and how to specify the PMM script file. After configuring those values you just need to start a PMM strategy.

**Comment by PtrckM** [comment]
Issue reported for non existing script are now fixed! LGTM

> * One minor issue when trying to run a non-existing script file (or accidentally entering a non-existing script file), the bot recognizes the file and still tries to start the strategy.
>

File: test/hummingbot/strategy/test_script_strategy_base.py (hunk 1 above)
Comment (aarmoa) [reply test/hummingbot/strategy/test_script_strategy_base.py]: Change applied

**Review by petioptrv — APPROVED** [review]
Changes LGTM!

**Comment by aarmoa** [comment]
Closing this PR after creating the official one in Hummingbot Foundation repository https://github.com/hummingbot/hummingbot/pull/5221

**Pull request closed without merging**